### PR TITLE
graph-builder: switch to using dkregistry and implement authenticated requests

### PR DESCRIPTION
**Pull request by steveej**

### CODE REVIEW

File: graph-builder/Cargo.toml [hunk 1]
```diff
@@ -18,3 +18,6 @@ serde_derive = "^1.0.70"
 serde_json = "^1.0.22"
 structopt = "^0.2.10"
 tar = "^0.4.16"
+dkregistry = { path = "/home/steveej/src/github/camallo/dkregistry-rs" }
```
Comment (steveej) [inline graph-builder/Cargo.toml]: # FIXME

File: docs/design/cincinnati.md [hunk 2]
```diff
@@ -39,6 +39,7 @@ An example of this metadata document can be seen here:
 
 ```json
 {
+	"kind": "cincinnati-metadata-v0",
```
Comment (steveej) [inline docs/design/cincinnati.md]: @crawford does this belong here? The text above says it's required

File: graph-builder/src/registry.rs [hunk 3]
```diff
@@ -37,23 +51,99 @@ impl Into<cincinnati::Release> for Release {
...
+    image: &str,
+    credentials_file: &str,
+) -> Result<RegistryClient, Error> {
+    let mut registry_client = dkregistry::v2::Client::configure(&mut tcore.handle())
+        .registry(host)
+        .insecure_registry(false)
+        .read_credentials(File::open(if credentials_file != "" {
+            credentials_file
+        } else {
+            "/dev/null"
```
Comment (steveej) [inline graph-builder/src/registry.rs]: # FIXME

File: graph-builder/src/registry.rs [hunk 4]
```diff
@@ -123,23 +254,13 @@ fn fetch_metadata(registry: &str, repo: &str, tag: &str) -> Result<release::Meta
...
+    trace!("Looking for {} in archive", metadata_filename);
 
-    ensure!(
-        response.status().is_success(),
-        "failed to fetch metadata document: {}",
-        response.status()
-    );
-
-    let mut archive = Archive::new(GzDecoder::new(response));
+    let mut archive = Archive::new(GzDecoder::new(&blob[..]));
```
Comment (steveej) [inline graph-builder/src/registry.rs]: @crawford @lucab can one of you confirm that this doesn't unnecessarily copy the whole blob?

File: docs/design/cincinnati.md (hunk 2 above)
Comment (lucab) [reply docs/design/cincinnati.md]: I think the `schema` integer became the type-tag `kind` on some later iteration of this design.

File: graph-builder/src/registry.rs (hunk 4 above)
Comment (lucab) [reply graph-builder/src/registry.rs]: I think it doesn't. You can also use a `Cursor` here.

Comment (crawford) [reply graph-builder/src/registry.rs]: I'm not sure what `blob` is, but that usually won't make a copy. You are asking for a slice with `[..]`, but because the size of a slice isn't known at compilation, you need to take a reference to it with `&`.

File: docs/design/cincinnati.md (hunk 2 above)
Comment (steveej) [reply docs/design/cincinnati.md]: It's required as it won't deserialize without it: 

https://github.com/openshift/cincinnati/blob/95db15a823b231636c7a3aaeef25b6c22e6c9a8b/graph-builder/src/release.rs#L22

https://github.com/openshift/cincinnati/blob/95db15a823b231636c7a3aaeef25b6c22e6c9a8b/graph-builder/src/release.rs#L47-L50

File: graph-builder/src/registry.rs (hunk 4 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: Here `blob` is a `Vec<u8>` which doesn't implement `Read` itself so I made it into a slice.

Comment (crawford) [reply graph-builder/src/registry.rs]: In that case, you can do

```rust
    let mut archive = Archive::new(GzDecoder::new(blob.as_ref()));
```

File: graph-builder/src/main.rs [hunk 5]
```diff
@@ -54,8 +54,7 @@ fn main() -> Result<(), Error> {
                 2 => LevelFilter::Debug,
                 _ => LevelFilter::Trace,
             },
-        )
```
Comment (crawford) [inline graph-builder/src/main.rs]: Don't accept the `rustfmt` suggestions from stable. It suffers from a [formatting bug](https://github.com/rust-lang-nursery/rustfmt/issues/2932).

File: graph-builder/Cargo.toml [hunk 6]
```diff
@@ -4,7 +4,7 @@ version = "0.1.0"
 authors = ["Alex Crawford <crawford@redhat.com>"]
 
 [dependencies]
-actix-web = "^0.7.8"
+actix-web = "^0.7.12"
```
Comment (crawford) [inline graph-builder/Cargo.toml]: Why does this need to be bumped?

File: graph-builder/Cargo.toml [hunk 7]
```diff
@@ -18,3 +18,6 @@ serde_derive = "^1.0.70"
 serde_json = "^1.0.22"
 structopt = "^0.2.10"
 tar = "^0.4.16"
+dkregistry = { git = "https://github.com/camallo/dkregistry-rs.git", rev = "bacca3b7fb80edf7c38035d6f355672aa2536dd4"}
```
Comment (crawford) [inline graph-builder/Cargo.toml]: Can you break this out into a separate commit?

File: graph-builder/src/registry.rs [hunk 8]
```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+type RegistryClient = dkregistry::v2::Client;
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Unless there is a good reason, I'd avoid type aliasing.

File: graph-builder/src/registry.rs [hunk 9]
```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
...
+
+fn registry_trim(src: &str) -> &str {
```
Comment (crawford) [inline graph-builder/src/registry.rs]: `trim_protocol` would be a more appropriate name.

File: graph-builder/src/registry.rs [hunk 10]
```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
...
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(
+    credentials_file: &str,
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Use a `&Path` instead. This parameter is also a path, not a file.

File: graph-builder/src/registry.rs [hunk 11]
```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
...
+) -> Result<Box<dyn std::io::Read>, Error> {
+    if credentials_file != "" {
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Use an `Option` instead of special-casing an empty string. This entire function can be reduced to:

```rust
  Ok(Box::new(credentials_path.map_or(std::io::empty(), |path| File::open(path)?)))
```

It might be easier to get rid of this function completely since the above expression is so simple. It would also avoid the boxing.

File: graph-builder/src/registry.rs [hunk 12]
```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
...
+    host: &str,
+    image: &str,
+    credentials_file: &str,
+) -> Result<RegistryClient, Error> {
+    let mut registry_client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(host)
+        .insecure_registry(false)
+        .read_credentials(get_credentials_or_empty_reader(credentials_file)?)
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
```
Comment (crawford) [inline graph-builder/src/registry.rs]: This shouldn't be necessary.

File: graph-builder/src/registry.rs [hunk 13]
```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
...
+    let logged_in = tcore
+        .run(fut_check_auth_prelogin)
+        .map_err(|e| format_err!("{}", e))?;
+    if logged_in {
+        return Err(format_err!("no login performed, but already authenticated"));
+    }
+
+    let fut_login = registry_client
+        .login(vec![&format!("repository:{}:pull", image)])
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Does this method require a Vec? I'd avoid heap allocations if possible.

File: graph-builder/src/registry.rs [hunk 14]
```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
...
+    if !tcore.run(fut_auth).map_err(|e| format_err!("{}", e))? {
+        return Err(format_err!("login failed"));
+    }
+    registry_client.set_token(Some(token.token()));
+    debug!("successfully logged in to {}", host);
+
+    let fut_check_v2_postlogin = registry_client
+        .is_v2_supported()
+        .map_err(|e| format_err!("{}", e))?;
+    if !tcore
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Instead of evaluating each of these futures independently, just chain them together and invoke them once.

File: graph-builder/src/registry.rs [hunk 15]
```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
...
-            ),
-            metadata: fetch_metadata(registry, repo, &tag)?,
-        })
+    let mut tcore = Core::new()?;
+
+    let registry_host = registry_trim(registry);
+    let registry_client =
+        get_authenticated_client(&mut tcore, registry_host, repo, credentials_file)?;
+
+    for tag in fetch_tags(&mut tcore, &registry_client, repo)? {
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Try something like this instead:

```rust
  fetch_tags(&mut tcore, &registry_client, repo)?
    .iter()
    .map(|tag| Release { source: ..., metadata: ... })
    .for_each(|release| {
      trace!("Found release {:?}", release);
      metadata.push(release);
    })
```

File: graph-builder/src/registry.rs [hunk 16]
```diff
@@ -63,21 +157,21 @@ struct Tags {
...
-        ensure!(
-            response.status().is_success(),
-            "failed to fetch image tags: {}",
-            response.status()
-        );
-
-        serde_json::from_str(&response.text()?)?
-    };
+fn fetch_tags(
+    tcore: &mut Core,
```
Comment (crawford) [inline graph-builder/src/registry.rs]: No need to evaluate the future here, just return it. You won't need to pass in the core.

**Review by crawford — CHANGES_REQUESTED** [review]

File: graph-builder/src/registry.rs (hunk 13 above)
Comment (lucab) [reply graph-builder/src/registry.rs]: Nope, I was just sloppy in the original API. Current master already takes a slice.

Comment (steveej) [reply graph-builder/src/registry.rs]: I've got that fixed locally after a clippy run. I'll push in a bit to update this PR

File: graph-builder/src/main.rs (hunk 5 above)
Comment (steveej) [reply graph-builder/src/main.rs]: Okay, odd enough ;-) I'll remember this when setting up the CI checks.

File: graph-builder/Cargo.toml (hunk 6 above)
Comment (steveej) [reply graph-builder/Cargo.toml]: It doesn't, I was experimenting with transitive dependencies at one point. Should I roll back?

File: graph-builder/Cargo.toml (hunk 7 above)
Comment (steveej) [reply graph-builder/Cargo.toml]: I can, I just need to put that commit first then, otherwise the code is broken. The commit containing this, and just this, will then have an unneeded dependency. Are we okay with having single commits in a PR which aren't consistent in itself?

File: graph-builder/src/registry.rs (hunk 12 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: Unfortunately there's no obvious better way to solve https://github.com/rust-lang-nursery/error-chain/issues/240.

File: graph-builder/src/registry.rs (hunk 15 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: Are you suggesting that for idiomatic reasons?

File: graph-builder/Cargo.toml (hunk 6 above)
Comment (crawford) [reply graph-builder/Cargo.toml]: Yes. Or update all of the dependencies in another commit.

File: graph-builder/Cargo.toml (hunk 7 above)
Comment (crawford) [reply graph-builder/Cargo.toml]: Of course. Commits within a PR are logically related.

File: graph-builder/src/registry.rs (hunk 15 above)
Comment (crawford) [reply graph-builder/src/registry.rs]: Yep. Alternatively, you can do:

```rust
  for release in fetch_tags(&mut tcore, &registry_client, repo)?
    .iter()
    .map(|tag| Release { source: ..., metadata: ... })
  {
    trace!("Found release {:?}", release);
    metadata.push(release);
  }
```

That might be easier to follow.

File: graph-builder/src/registry.rs [hunk 17]
```diff
@@ -37,24 +46,164 @@ impl Into<cincinnati::Release> for Release {
...
+        Ok(Box::new(std::io::empty()))
+    }
+}
+
+pub fn authenticate_client<'a>(
+    client: &'a mut dkregistry::v2::Client,
+    login_scope: &'a str,
+) -> impl futures::future::Future<Item = &'a dkregistry::v2::Client, Error = dkregistry::errors::Error>
+{
+    futures::future::ok(client)
```
Comment (crawford) [inline graph-builder/src/registry.rs]: This isn't needed. You can start the future chain with the call to `dclient.is_v2_supported()` below.

File: graph-builder/src/registry.rs (hunk 11 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: If I write the function as suggested:

```rust
fn get_credentials_or_empty_reader(
    credentials_path: &Path,
) -> Result<Box<dyn std::io::Read>, Error> {
    Ok(Box::new(
        credentials_path.map_or(std::io::empty(), |path| File::open(path)?),
    ))
}
```

the compiler errors out:


```
   Compiling graph-builder v0.1.0 (/home/steveej/src/github/openshift/cincinnati/graph-builder)                                              
error[E0599]: no method named `map_or` found for type `&std::path::Path` in the current scope                                                
  --> graph-builder/src/registry.rs:58:26                                                                                                    
   |                                                                                                                                         
58 |         credentials_path.map_or(std::io::empty(), |path| File::open(path)?),                                                            
   |                          ^^^^^^                                                                                                         
                                                                                                                                             
error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
  --> graph-builder/src/registry.rs:58:58                                                                                                    
   |                                                                                                                                         
58 |         credentials_path.map_or(std::io::empty(), |path| File::open(path)?),                                                            
   |                                                          ^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `std::fs::File`
   |                                                                                                                                         
   = help: the trait `std::ops::Try` is not implemented for `std::fs::File`                                                                  
   = note: required by `std::ops::Try::from_error`                                                                                           
                                                                                                                                             
error[E0277]: the size for values of type `dyn std::io::Read` cannot be known at compilation time                                            
  --> graph-builder/src/registry.rs:57:8                                                                                                     
   |                                                                                                                                         
57 |     Ok(Box::new(                                                                                                                        
   |        ^^^^^^^^ doesn't have a size known at compile-time                                                                               
   |                                                                                                                                         
   = help: the trait `std::marker::Sized` is not implemented for `dyn std::io::Read`                                                         
   = note: to learn more, visit <https://doc.rust-lang.org/book/second-edition/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
   = note: required by `<std::boxed::Box<T>>::new`                                                                                           
                                                                                                                                             
error: aborting due to 3 previous errors                                                                                                     
                                                                                                                                             
Some errors occurred: E0277, E0599.                                                                                                          
For more information about an error, try `rustc --explain E0277`.                                                                            
error: Could not compile `graph-builder`.                                                                                                    

To learn more, run the command again with --verbose.
error: Recipe `run-graph-builder` failed on line 15 with exit code 101
```

If your motivation is simplicity then I suggest we leave it as is. I initially tried to use `Path` for this and ran into complications that didn't seem worth the effort. I'll take another closer look tomorrow.

Comment (steveej) [reply graph-builder/src/registry.rs]: I went for `Option<PathBuf>` since that's [officially supported by structopt](https://docs.rs/structopt/0.2.12/structopt/?search=#how-to-derivestructopt) and was the least effort. I couldn't get around using a separate function returning a the trait object for declaring the implementation of `std::io::Read`.

File: graph-builder/src/registry.rs [hunk 18]
```diff
@@ -37,24 +46,164 @@ impl Into<cincinnati::Release> for Release {
...
+    let authenticated_client = tcore
+        .run(authenticate_client(&mut client, &login_scope))
+        .map_err(|e| format_err!("{}", e))?;
+
+    let releases_future = authenticated_client
+        .get_tags(&repo, None)
+        .and_then(|tag| {
+            trace!("processing: {}:{}", &repo, &tag);
+
+            let tag_clone0 = tag.clone(); // TODO(steveeJ): is there a way to avoid this?
```
Comment (crawford) [inline graph-builder/src/registry.rs]: I _think_ NLL will solve this for you. Switching over to Rust 2018 might do the trick.

File: graph-builder/src/registry.rs [hunk 19]
```diff
@@ -37,24 +46,164 @@ impl Into<cincinnati::Release> for Release {
...
+    let registry_host = trim_protocol(registry);
+
+    let mut client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(registry_host)
+        .insecure_registry(false)
+        .read_credentials(get_credentials_or_empty_reader(credentials_path))
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let login_scope = format!("repository:{}:pull", &repo);
```
Comment (crawford) [inline graph-builder/src/registry.rs]: I prefer not to use intermediate variables unless it helps with readability. In this case, I don't think it buys you much so I would probably inline this.

File: graph-builder/src/registry.rs [hunk 20]
```diff
@@ -37,24 +46,164 @@ impl Into<cincinnati::Release> for Release {
...
+
+    let authenticated_client = tcore
+        .run(authenticate_client(&mut client, &login_scope))
+        .map_err(|e| format_err!("{}", e))?;
+
+    let releases_future = authenticated_client
```
Comment (crawford) [inline graph-builder/src/registry.rs]: I don't think we need to use Hungarian Notation. I would just call this `releases`.

File: graph-builder/src/registry.rs [hunk 21]
```diff
@@ -37,24 +46,164 @@ impl Into<cincinnati::Release> for Release {
...
+                        let release = Release {
+                            source: format!("{}/{}:{}", &registry_host, &repo, &tag),
+                            metadata: metadata,
+                        };
+                        trace!("Found release '{:?}'", release);
+                        Ok(release)
+                    })
+            });
+
+            let futures: Vec<_> = layer_digests_mapped_to_releases.collect();
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Use turbofish to avoid the extraneous assignment:

```Rust
Ok(layer_digests_mapped_to_releases.collect::<Vec<_>>())
```

**Review by crawford — CHANGES_REQUESTED** [review]

File: graph-builder/src/registry.rs (hunk 20 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: I think it's immensely helpful to distinguish between the future version and the concrete version. If you look closely I've used this notation throughout the whole file. @lucab any opinion on this?

File: graph-builder/src/registry.rs (hunk 18 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: I'm building with stable as of now. Should we swtich over to nightly until 2018 is released?

File: graph-builder/src/registry.rs (hunk 21 above)
Comment (crawford) [reply graph-builder/src/registry.rs]: We should favor readability. That's why I made that suggestion ;)  Otherwise you quickly end up with something that looks like Go, where you read 20 lines of code to figure out that it is doing a trivial operation.

File: graph-builder/src/registry.rs (hunk 20 above)
Comment (crawford) [reply graph-builder/src/registry.rs]: If we are going all in with Futures, there shouldn't be a concrete and future version; there should just be futures (except for that one value at the end).

File: graph-builder/src/registry.rs (hunk 19 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: You can discuss this with the compiler

```
error[E0597]: borrowed value does not live long enough                                                                                       
   --> graph-builder/src/registry.rs:125:14                                                                                                  
    |                                                                                                                                        
125 |             &format!("repository:{}:pull", &repo),                                                                                     
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ temporary value does not live long enough                                            
126 |         ))                                                                                                                             
127 |         .map_err(|e| format_err!("{}", e))?;                                                                                           
    |                                            - temporary value dropped here while still borrowed                                         
...                                                                                                                                          
208 | }                                                                                                                                      
    | - temporary value needs to live until here                                                                                             
    |                                                                                                                                        
    = note: consider using a `let` binding to increase its lifetime                                                                          
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
                                                                                                                                             
error: aborting due to previous error                                                                                                        
                                                                                                                                             
For more information about this error, try `rustc --explain E0597`.                                                                          
error: Could not compile `graph-builder`.  
```

File: graph-builder/src/registry.rs [hunk 22]
```diff
@@ -37,24 +46,163 @@ impl Into<cincinnati::Release> for Release {
...
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(credentials_path: &Option<PathBuf>) -> Box<dyn std::io::Read> {
+    let empty = Box::new(std::io::empty());
+    match credentials_path {
+        Some(path) => match File::open(path) {
+            Ok(file) => Box::new(file),
+            _ => empty,
```
Comment (crawford) [inline graph-builder/src/registry.rs]: We shouldn't eat this error.

File: graph-builder/src/registry.rs [hunk 23]
```diff
@@ -37,24 +46,163 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+fn trim_protocol(src: &str) -> &str {
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(credentials_path: &Option<PathBuf>) -> Box<dyn std::io::Read> {
+    let empty = Box::new(std::io::empty());
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Don't do heap allocations unless it's actually necessary.

File: graph-builder/src/registry.rs [hunk 24]
```diff
@@ -37,24 +46,163 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+fn trim_protocol(src: &str) -> &str {
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(credentials_path: &Option<PathBuf>) -> Box<dyn std::io::Read> {
```
Comment (crawford) [inline graph-builder/src/registry.rs]: If you just inline this function, you can avoid the heap allocation. I don't think this helps much with readability.

File: graph-builder/src/registry.rs [hunk 25]
```diff
@@ -37,24 +46,163 @@ impl Into<cincinnati::Release> for Release {
...
+                });
+
+            let manifest_future = authenticated_client.get_manifest(&repo, &tag);
+
+            let layer_digests_future =
+                manifest_kind_future
+                    .join(manifest_future)
+                    .map(|(manifest_kind, manifest)| match manifest_kind {
+                        dkregistry::mediatypes::MediaTypes::ManifestV2S1Signed => {
+                            let m: dkregistry::v2::manifest::ManifestSchema1Signed =
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Is this intermediate variable necessary?

File: graph-builder/src/registry.rs [hunk 26]
```diff
@@ -37,24 +46,163 @@ impl Into<cincinnati::Release> for Release {
...
         })
-    }
-    Ok(metadata)
+        .map_err(|e| format_err!("{}", e))
+        .and_then(|layer_digests_tag| {
+            let (layer_digests, tag) = layer_digests_tag?;
+            trace!("tag: {:?} layer_digests: {:?}", &tag, &layer_digests);
+
+            let layer_digests_mapped_to_releases = layer_digests.iter().map(|layer_digest| {
+                let (registry_host, repo, tag) = (registry_host.clone(), repo.clone(), tag.clone());
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Are these clones actually necessary?

File: graph-builder/src/registry.rs [hunk 27]
```diff
@@ -37,24 +46,163 @@ impl Into<cincinnati::Release> for Release {
...
+                        let release = Release {
+                            source: format!("{}/{}:{}", &registry_host, &repo, &tag),
+                            metadata: metadata,
+                        };
+                        trace!("Found release '{:?}'", release);
+                        Ok(release)
+                    })
+            });
+
+            Ok(layer_digests_mapped_to_releases.collect::<Vec<_>>())
```
Comment (crawford) [inline graph-builder/src/registry.rs]: We don't want to collect the releases. We just want the first one we find.

**Review by crawford — CHANGES_REQUESTED** [review]

File: graph-builder/src/registry.rs (hunk 26 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: Yes, because they are consumed by the closures

File: graph-builder/src/registry.rs (hunk 24 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: Using `Box<dyn std::io::Read>` as the return type is the only way I've found to tell the compiler that all cases yield a reader. I've double-checked inlining this yesterday and couldn't make it work. Do you want me to triple-check?

File: graph-builder/src/registry.rs (hunk 25 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: It seems to be. When I rewrite one of the match arms as 

```rust
                        dkregistry::mediatypes::MediaTypes::ManifestV2S1Signed => Ok((
                            serde_json::from_slice(manifest.as_slice())?.get_layers(),
                            tag,
                        )),
```
the compiler prints:
```console
error[E0282]: type annotations needed                                                                                                        
   --> graph-builder/src/registry.rs:151:29                                                                                                  
    |                                                                                                                                        
151 |                             serde_json::from_slice(manifest.as_slice())?.get_layers(),                                                 
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type                                             
    |                                                                                                                                        
    = note: type must be known at this point                                                                                                 
                                                                                                                                             
error: aborting due to previous error                                                                                                        
                                                                                                                                             
For more information about this error, try `rustc --explain E0282`. 
```

File: graph-builder/src/registry.rs (hunk 27 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: This collects to a vector of futures which are part of the stream. I had that information in the variable name you had me remove in the previous iteration.

Comment (steveej) [reply graph-builder/src/registry.rs]: On second thought I noticed that the `select_ok` which is further down likely creates a race. I will double-check this, as the resolution to that will resolve this issue too.

Comment (steveej) [reply graph-builder/src/registry.rs]: The suspicion proved correct and I was able to reproduce the bug in a test over in #8. I implemented a simple fix in https://github.com/openshift/cincinnati/pull/4/commits/0121a299df2306f699ec3bdf6cfcb74150fbdc0f.

File: graph-builder/src/registry.rs (hunk 26 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: I was able to remove these in the split version of the future chain too.

File: graph-builder/src/registry.rs (hunk 25 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: I was able to get rid of the assignments by using the turbofish annotation at `from::slice`

File: graph-builder/src/registry.rs (hunk 24 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: I managed to inline the functionality yet the `Box` is still required. I suggest we leave it in and revisit on subsequent compiler releases.

File: graph-builder/src/registry.rs (hunk 18 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: This was resolved too by the future split.

File: graph-builder/src/registry.rs [hunk 28]
```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
...
+    let mut tcore = Core::new()?;
+
+    let registry_host = trim_protocol(registry);
+
+    let mut client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(registry_host)
+        .insecure_registry(false)
+        .read_credentials(|| -> Result<Box<dyn std::io::Read>, Error> {
+            match credentials_path {
+                Some(path) => match File::open(path) {
```
Comment (crawford) [inline graph-builder/src/registry.rs]: I don't think we should re-read this file every time we want to fetch the releases. Can we move this into main?

File: graph-builder/src/registry.rs [hunk 29]
```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
...
+        .collect();
+
+    let releases = tcore.run(releases)?;
+    Ok(releases)
+}
+
+fn get_layers_for_manifest_with_kind(
+    manifest_kind: Option<dkregistry::mediatypes::MediaTypes>,
+    manifest: Vec<u8>,
+    tag: String,
```
Comment (crawford) [inline graph-builder/src/registry.rs]: Can we remove this argument since it isn't needed?

File: graph-builder/src/registry.rs [hunk 30]
```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
...
+                tag,
+            )
+        })
+        .collect();
+
+    let releases = tcore.run(releases)?;
+    Ok(releases)
+}
+
+fn get_layers_for_manifest_with_kind(
```
Comment (crawford) [inline graph-builder/src/registry.rs]: This is a pretty verbose name. Maybe just `get_layers()`?

File: graph-builder/src/registry.rs [hunk 31]
```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
...
+                .join(authenticated_client.get_manifest(&repo, &tag))
+                .map(|(manifest_kind, manifest)| {
+                    get_layers_for_manifest_with_kind(manifest_kind, manifest, tag)
+                })
+                .map_err(|e| format_err!("{}", e))
+        })
+        .and_then(|tag_layer_digests| match tag_layer_digests {
+            Ok((tag, layer_digests)) => {
+                if layer_digests.len() == 0 {
+                    bail!("layer_digests list empty")
```
Comment (crawford) [inline graph-builder/src/registry.rs]: This doesn't seem like an error condition. We can handle this gracefully.

**Review by crawford — CHANGES_REQUESTED** [review]
I haven't made it all the way through this yet. `clippy` also has some suggestions you should follow.

File: graph-builder/src/registry.rs (hunk 28 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: The read happens within the library so if we want to avoid re-read we have to move the client creation to main. I don't think it's very expensive as it's certainly buffered after the first read.

File: graph-builder/src/registry.rs (hunk 29 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: It's passed through to the `Release { ... }` creation.

File: graph-builder/src/registry.rs (hunk 31 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: Setting the context for the discussion: this is the case when there are no layers available for a discovered tag, which seems highly unexpected. Do you think it's better to just log the errors and drive the whole chain to completion and yield an empty result vector?

File: graph-builder/src/registry.rs [hunk 32]
```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
...
+            // the tags should be ordered lexically but they aren't
+            .get_tags(&repo, Some(20))
+            .map_err(|e| format_err!("{}", e))
+            .collect(),
+    ) {
+        Ok(mut tags) => {
+            if tags.len() == 0 {
+                bail!("{}/{} has no tags", registry, repo);
+            }
+            tags.sort();
```
[the remark this answers is not in the document]
Comment (steveej) [reply graph-builder/src/registry.rs:133]: I designed the tests in a way to expect a certain vector of results because the documentation says the tags are returned in a lexical sorted order. Quay seems to diverge from that so this sort enforces what the docker registry API documents.

File: graph-builder/src/registry.rs [hunk 33]
```diff
@@ -37,24 +45,231 @@ impl Into<cincinnati::Release> for Release {
...
+        authenticated_client
+            // According to https://docs.docker.com/registry/spec/api/#listing-image-tags
+            // the tags should be ordered lexically but they aren't
+            .get_tags(&repo, Some(20))
+            .map_err(|e| format_err!("{}", e))
+            .collect(),
+    ) {
+        Ok(mut tags) => {
+            if tags.len() == 0 {
+                bail!("{}/{} has no tags", registry, repo);
```
[the remark this answers is not in the document]
Comment (steveej) [reply graph-builder/src/registry.rs]: Please consider this in the answer to https://github.com/openshift/cincinnati/pull/4/files/835a3685c6c29f76bb999b96ec72f4ce00b37550#r233220468

File: graph-builder/src/registry.rs (hunk 31 above)
Comment (crawford) [reply graph-builder/src/registry.rs]: Yes. Logging the warning and then just returning an empty vector is fine.

File: graph-builder/src/registry.rs (hunk 32 above)
[the remark this answers is not in the document]
Comment (crawford) [reply graph-builder/src/registry.rs:133]: We should really try to avoid coupling to Quay as much as possible. CI cannot depend on it being up and they have already expressed concern over the amount of load we are putting on them. I'd rather we just adjusted our tests instead of the product.

File: graph-builder/src/registry.rs (hunk 28 above)
Comment (crawford) [reply graph-builder/src/registry.rs]: I think this is okay for now, but I think we should eventually move it.

File: graph-builder/src/registry.rs (hunk 29 above)
Comment (crawford) [reply graph-builder/src/registry.rs]: I don't follow. I was able to remove this in my test branch and create the tuple up at the call site. That seems like the more appropriate place to inject the tag.

Comment (steveej) [reply graph-builder/src/registry.rs]: That works too

File: graph-builder/src/registry.rs (hunk 30 above)
Comment (steveej) [reply graph-builder/src/registry.rs]: Let's have a compromise and use `get_layer_digests` which is accurate about the return value, the rest is indeed explained via the input arguments.

File: graph-builder/src/registry.rs (hunk 32 above)
[the remark this answers is not in the document]
Comment (steveej) [reply graph-builder/src/registry.rs:133]: It's possible to neglect the order in the tests, with the downside of making a failed test more difficult to debug.

**Review by crawford — APPROVED** [review]
